programs/castle-vault/src/instructions/sell_port_reward.rs

```diff
@@ -0,0 +1,142 @@
+use anchor_lang::prelude::*;
+use anchor_lang::solana_program::program::invoke_signed;
+use anchor_spl::token::{Token, TokenAccount};
+
+use crate::{
+    errors::ErrorCode,
+    state::{DexStates, OrcaLegacyAccounts, Vault, VaultPortAdditionalState},
+};
+
+#[derive(Accounts)]
+pub struct SellPortReward<'info> {
+    /// Vault state account
+    /// Checks that the accounts passed in are correct
+    #[account(
+        mut,
+        has_one = vault_authority,
+        has_one = vault_reserve_token
+    )]
+    pub vault: Box<Account<'info, Vault>>,
+
+    /// Authority that the vault uses for lp token mints/burns and transfers to/from downstream assets
+    /// CHECK: safe
+    pub vault_authority: AccountInfo<'info>,
+
+    #[account(
+        seeds = [vault.key().as_ref(), b"port_additional_state".as_ref()], 
+        bump
+    )]
+    pub port_additional_states: Box<Account<'info, VaultPortAdditionalState>>,
+
+    #[account(
+        seeds = [vault.key().as_ref(), b"dex_states".as_ref()], 
+        bump
+    )]
+    pub dex_states: Box<Account<'info, DexStates>>,
+
+    #[account(
+        seeds = [vault.key().as_ref(), b"dex_orca_legacy".as_ref()],
+        bump,
+        has_one = orca_swap_program,
+    )]
+    pub orca_legacy_accounts: Box<Account<'info, OrcaLegacyAccounts>>,
+
+    /// CHECK: safe
+    pub orca_swap_state: AccountInfo<'info>,
+
+    // DANGER why can we ignore this? because the the Orca program will check this
+    //        and fail the CPI if this account is invalid.
+    // TODO Security audit to ensure it's really ok.
+    /// CHECK: safe
+    //#[soteria(ignore)]
+    pub orca_swap_authority: AccountInfo<'info>,
+
+    /// CHECK: safe
+    //#[soteria(ignore)]
+    #[account(mut)]
+    pub orca_input_token_account: AccountInfo<'info>,
+
+    /// CHECK: safe
+    //#[soteria(ignore)]
+    #[account(mut)]
+    pub orca_output_token_account: AccountInfo<'info>,
+
+    /// CHECK: safe
+    //#[soteria(ignore)]
+    #[account(mut)]
+    pub orca_swap_token_mint: AccountInfo<'info>,
+
+    /// CHECK: safe
+    //#[soteria(ignore)]
+    #[account(mut)]
+    pub orca_fee_account: AccountInfo<'info>,
+
+    /// CHECK: safe
+    #[account(executable)]
+    pub orca_swap_program: AccountInfo<'info>,
```

Comment:
Need to verify that this address is correct

Reply:
It's verified at line 40 of this file, `has_one = orca_swap_program`.
`orca_legacy_accounts` is verified in the handler.

It looks a bit indirect, the main reason for that is orca does not have an their own rust sdk like port, which we can use to get the expected program ID. We can either hard-code the program ID in our program or do what I did, which is to store it as a pubkey and verify using `has_one`

Would love to explore other alternatives.